### PR TITLE
Adding toCSS method WIP - RFC

### PR DESCRIPTION
**Pull request by fskreuz**
Exposes the CSS generated by Ractive via a new toCSS method, plus a few cleanups and tests here and there. Since there wasn't a PR (just a commit on some fork), decided to create one.

ToDo:

- [x] Render CSS  https://github.com/ractivejs/ractive/issues/1127
- [x] Multiple copies of Ractive generating CSS https://github.com/ractivejs/ractive/issues/2292
- [x] Warn when using `css` option for instances https://github.com/ractivejs/ractive/issues/2365
- [x] `toCSS` to trace through instance components only https://github.com/ractivejs/ractive/pull/2362#issuecomment-179639832

### CODE REVIEW

**Comment by fskreuz** [comment]
@Rich-Harris I remember components had uuid as identifiers. Why the switch to numbers?

**Comment by fskreuz** [comment]
Tracking https://github.com/ractivejs/ractive/issues/395

**Comment by evs-chris** [comment]
It looks like this would also fix #2292, since the uid goes from monotonic to uuid. Or have I misread that entirely?

Does this also cover the case of adding css using `new Ractive({ ... css: ... })` and `new SomeComponent({ ... css: ... })`?

I haven't really made use of component css up to this point, but this looks like it should do nicely. There are a few minor formatting inconsistencies with spacing around ifs and method defs. Sorry. I'm the whitespace version of a grammar nazi, even though the Ractive style-guide doesn't really agree with my personal aesthetics.

**Comment by fskreuz** [comment]
> It looks like this would also fix #2292, since the uid goes from monotonic to uuid. Or have I misread that entirely?

Yep. Using uuid should prevent two copies Ractive from colliding scoping ids. It's just that I don't know how to test two instances of Ractive on one page/server.

> Does this also cover the case of adding css using new Ractive({ ... css: ... }) and new SomeComponent({ ... css: ... })?

This currently only works with `Ractive.extend`. I asked about that in https://github.com/ractivejs/ractive/issues/2365 if it was intended behavior for it to only work on `Ractive.extend` or if it should work on both cases. If it should work on both, then there's more work to be done. :smile: 

> There are a few minor formatting inconsistencies with spacing around ifs and method defs.

Yep, working on my editor to do that for me. ESLint and editorconfig are quite new to me. :smile:

**Comment by martypdx** [comment]
> This currently only works with Ractive.extend. I asked about that in #2365 if it was intended behavior for it to only work on Ractive.extend or if it should work on both cases

Added a comment on that issue.

> Yep. Using uuid should prevent two copies Ractive from colliding scoping ids. It's just that I don't know how to test two instances of Ractive on one page/server on the unit tests.

Add two divs to fixture and assign them to respective instance:

```
var div1 = document.createElement('div');
var div2 = document.createElement('div');
fixture.appendChild( div1 );
fixture.appendChild( div2 );

var r1 = new Ractive({ el: div1 });
var r2 = new Ractive({ el: div2 });
```

**Comment by martypdx** [comment]
BTW - UUID's where removed from instances because with large numbers of component instances becomes costly. Totally fine on per Component definition.

**Comment by fskreuz** [comment]
> UUID's where removed from instances because with large numbers of component instances becomes costly

Ahh, makes sense. 

> Add two divs to fixture and assign them to respective instance

That's a bit different from https://github.com/ractivejs/ractive/issues/2292 where the case is about multiple copies of Ractive, each recording their own ids, and possibly colliding because they're numbers. :smile:

**Comment by fskreuz** [comment]
Updated to test isolated instances of Ractive by loading them up in iframes and building components with them. Appveyor failing though.

**Comment by evs-chris** [comment]
> Appveyor failing though.

Yeah... every now and then the AppVeyor CI will decide to work. We mostly just go by whatever Travis has to say at this point. :smile:

File: src/global/css.js [hunk 1]
```diff
@@ -1,62 +1,53 @@
...
+let styleSheet = null;
+let styleProperty = null;
+
+function add( styleDefinition ) {
+	styleDefinitions.push( styleDefinition );
+	isDirty = true;
+}
+
+function toCSS() {
+	return PREFIX + styleDefinitions.map( s => `\n/* {${s.id}} */\n${s.styles}` ).join( '\n' );
```
Comment (fskreuz) [inline src/global/css.js]: This implementation of `toCSS` (as a prototype method) doesn't seem to make sense when this `toCSS` returns all the styles defined. So...

- Either make `toCSS` a static function that returns all CSS defined in Ractive.
- Gather all the IDs from the instance down to the descendants, and grab only their CSS.

**Comment by evs-chris** [comment]
In the context of a long running server process that is potentially munging lots of Ractive instances, I'd go with the latter. I think that's one of the primary use cases for this, right?

**Comment by evs-chris** [comment]
@fskreuz this has to be the most tested PR I've seen :smile:. Is it ready to go?

The only question I came up with whilst perusing the source was if it would be possible to handle the `toHTML()` case server-side for `enhance`ment purposes, but I couldn't think of a way to do so.

**Comment by fskreuz** [comment]
> this has to be the most tested PR I've seen

Taking no chances :D

> Is it ready to go?

Looks ok to me. Just waiting for eyes to take a look at it and say it's good.

> if it would be possible to handle the toHTML() case server-side for enhancement purposes

If you mean merge this to the `toHTML()` method or something, I'll take a look.

**Comment by martypdx** [comment]
> If you mean merge this to the toHTML() method or don't re-apply when existing, I'll take a look.

What's the thought on this? that `toHTML()` should include a `style` with the css? Or should we just leave that to the caller? ``` `<style>${ractive.toCss()}</style>${ractive.toHtml()}` ```

Everything else LGTM. :+1:

**Comment by evs-chris** [comment]
I would just leave it to the caller. Was just wondering if there was an obvious way to handle it together that I couldn't noodle out.

**Comment by martypdx** [comment]
done like dinner

**Comment by fskreuz** [comment]
:+1:

**Pull request merged**